File: app/src/main/java/com/teambrella/android/ui/chat/ChatFragment.java
package com.teambrella.android.ui.chat;

import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v7.widget.LinearLayoutManager;
import android.text.Html;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.api.server.TeambrellaUris;
import com.teambrella.android.data.base.IDataPager;
import com.teambrella.android.image.glide.GlideApp;
import com.teambrella.android.ui.TeambrellaUser;
import com.teambrella.android.ui.base.ADataFragment;
import com.teambrella.android.ui.base.ADataPagerProgressFragment;
import com.teambrella.android.ui.base.ATeambrellaDataPagerAdapter;
import com.teambrella.android.ui.claim.ClaimActivity;
import com.teambrella.android.ui.claim.ClaimVotingFragment;
import com.teambrella.android.ui.teammate.TeammateActivity;

import java.util.Locale;

import io.reactivex.Notification;
import io.reactivex.Observable;


/**
 * Claim Chat Fragment
 */
public class ChatFragment extends ADataPagerProgressFragment<IChatActivity> {

    private static final String VOTING_FRAGMENT_TAG = "voting_fragment_tag";


    Long mLastRead = null;


    @SuppressWarnings("FieldCanBeLocal")
    private View mVotingPanelView;
    private TextView mTitleView;
    private TextView mSubtitleView;
    private TextView mVoteValueView;
    private TextView mVoteTitleView;
    private View mVoteButton;
    private ImageView mIcon;
    private String mUserName;

    @Override
    protected ATeambrellaDataPagerAdapter getAdapter() {
        int mode = ChatAdapter.MODE_DISCUSSION;
        switch (TeambrellaUris.sUriMatcher.match(mDataHost.getChatUri())) {
            case TeambrellaUris.CLAIMS_CHAT:
                mode = ChatAdapter.MODE_CLAIM;
                break;
            case TeambrellaUris.TEAMMATE_CHAT:
                mode = ChatAdapter.MODE_APPLICATION;
                break;
            case TeambrellaUris.CONVERSATION_CHAT:
                mode = ChatAdapter.MODE_CONVERSATION;
                break;
        }
        return new ChatAdapter(mDataHost.getPager(mTag), mDataHost.getTeamId(), mode, TeambrellaUser.get(getContext()).getUserId());
    }


    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setRefreshable(false);
        mList.setBackgroundColor(Color.TRANSPARENT);
        ((LinearLayoutManager) mList.getLayoutManager()).setStackFromEnd(true);

        mVotingPanelView = view.findViewById(R.id.voting_panel);
        mTitleView = view.findViewById(R.id.title);
        mSubtitleView = view.findViewById(R.id.subtitle);
        mVoteValueView = view.findViewById(R.id.vote_value);
        mVoteButton = view.findViewById(R.id.vote);
        mIcon = view.findViewById(R.id.image);
        mVoteTitleView = view.findViewById(R.id.your_vote_title);


        switch (TeambrellaUris.sUriMatcher.match(mDataHost.getChatUri())) {
            case TeambrellaUris.CLAIMS_CHAT:
                mVotingPanelView.setVisibility(View.VISIBLE);
                mVotingPanelView.setOnClickListener(this::onClaimClickListener);
                mVoteButton.setOnClickListener(this::onClaimClickListener);
                FragmentManager fragmentManager = getChildFragmentManager();
                if (fragmentManager.findFragmentByTag(VOTING_FRAGMENT_TAG) == null) {
                    fragmentManager.beginTransaction().add(R.id.voting_container,
                            ADataFragment.getInstance(new String[]{ChatActivity.CLAIM_DATA_TAG, ChatActivity.VOTE_DATA_TAG}
                                    , ClaimVotingFragment.class)
                            , VOTING_FRAGMENT_TAG)
                            .commit();
                }
                mDataHost.load(ChatActivity.CLAIM_DATA_TAG);
                break;
            case TeambrellaUris.TEAMMATE_CHAT:
                mVotingPanelView.setVisibility(View.VISIBLE);
                mVotingPanelView.setOnClickListener(this::onTeammateClickListener);
                mVoteButton.setOnClickListener(this::onTeammateClickListener);
                view.findViewById(R.id.voting_container).setVisibility(View.GONE);
                break;
            case TeambrellaUris.CONVERSATION_CHAT:
                mVotingPanelView.setVisibility(View.GONE);
                mList.setPadding(mList.getPaddingLeft(), 0, mList.getPaddingRight(), mList.getPaddingBottom());
                view.findViewById(R.id.voting_container).setVisibility(View.GONE);
                break;
            case TeambrellaUris.FEED_CHAT:
                mVotingPanelView.setVisibility(View.GONE);
                mList.setPadding(mList.getPaddingLeft(), 0, mList.getPaddingRight(), mList.getPaddingBottom());
                view.findViewById(R.id.voting_container).setVisibility(View.GONE);
                break;

        }
        mList.setItemAnimator(null);
    }


    @Override
    protected int getContentLayout() {
        return R.layout.fragment_chat;
    }

    @Override
    protected void onDataUpdated(Notification<JsonObject> notification) {
        super.onDataUpdated(notification);
        if (notification.isOnNext()) {
            JsonWrapper data = new JsonWrapper(notification.getValue()).getObject(TeambrellaModel.ATTR_DATA);
            JsonWrapper metadata = new JsonWrapper(notification.getValue()).getObject(TeambrellaModel.ATTR_METADATA_);
            if (metadata != null && (metadata.getBoolean(TeambrellaModel.ATTR_METADATA_FORCE, false)
                    || metadata.getBoolean(TeambrellaModel.ATTR_METADATA_RELOAD, false)) && metadata.getInt(TeambrellaModel.ATTR_METADATA_SIZE) > 0) {
                mList.getLayoutManager().scrollToPosition(mAdapter.getItemCount() - 1);
            }

            if (mLastRead == null) {
                JsonWrapper discussionPart = data.getObject(TeambrellaModel.ATTR_DATA_ONE_DISCUSSION);
                mLastRead = discussionPart.getLong(TeambrellaModel.ATTR_DATA_LAST_READ, -1L);
            }


            JsonWrapper basicPart = data.getObject(TeambrellaModel.ATTR_DATA_ONE_BASIC);
            JsonWrapper votingPart = data.getObject(TeambrellaModel.ATTR_DATA_ONE_VOTING);
            if (basicPart != null) {

                switch (TeambrellaUris.sUriMatcher.match(mDataHost.getChatUri())) {
                    case TeambrellaUris.CLAIMS_CHAT: {
                        Observable.fromArray(basicPart).map(json -> json.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO))
                                .map(uri -> GlideApp.with(this).load(getImageLoader().getImageUrl(uri)))
                                .subscribe(requestCreator -> requestCreator
                                        .apply(new RequestOptions().transforms(new CenterCrop()
                                                , new RoundedCorners(getResources().getDimensionPixelOffset(R.dimen.rounded_corners_4dp))).placeholder(R.drawable.picture_background_round_4dp))
                                        .into(mIcon), throwable -> {
                                });
                        mTitleView.setText(basicPart.getString(TeambrellaModel.ATTR_DATA_MODEL));
                        JsonWrapper teamPart = data.getObject(TeambrellaModel.ATTR_DATA_ONE_TEAM);
                        mSubtitleView.setText(Html.fromHtml(getString(R.string.claim_amount_format_string, Math.round(basicPart.getFloat(TeambrellaModel.ATTR_DATA_CLAIM_AMOUNT))
                                , teamPart != null ? teamPart.getString(TeambrellaModel.ATTR_DATA_CURRENCY) : "")));

                        if (votingPart == null) {
                            mVoteTitleView.setText(R.string.team_vote);
                            setClaimVoteValue(basicPart.getFloat(TeambrellaModel.ATTR_DATA_REIMBURSEMENT, -1f));
                        }
                    }
                    break;
                    case TeambrellaUris.TEAMMATE_CHAT: {
                        Observable.fromArray(basicPart).map(json -> json.getString(TeambrellaModel.ATTR_DATA_AVATAR))
                                .map(uri -> GlideApp.with(this).load(getImageLoader().getImageUrl(uri)))
                                .subscribe(requestCreator -> requestCreator.apply(RequestOptions.circleCropTransform())
                                        .into(mIcon), throwable -> {
                                    // 8)
                                });
                        mUserName = basicPart.getString(TeambrellaModel.ATTR_DATA_NAME);
                        mTitleView.setText(mUserName);
                        mSubtitleView.setText(getString(R.string.object_format_string
                                , basicPart.getString(TeambrellaModel.ATTR_DATA_MODEL)
                                , basicPart.getString(TeambrellaModel.ATTR_DATA_YEAR)));
                        mSubtitleView.setAllCaps(true);

                        if (votingPart == null) {
                            mVoteTitleView.setText(R.string.risk);
                            setTeammateVoteValue(basicPart.getFloat(TeambrellaModel.ATTR_DATA_RISK));
                        }
                    }
                    break;
                }
            }

            if (votingPart != null) {
                switch (TeambrellaUris.sUriMatcher.match(mDataHost.getChatUri())) {
                    case TeambrellaUris.CLAIMS_CHAT:
                        setClaimVoteValue(votingPart.getFloat(TeambrellaModel.ATTR_DATA_MY_VOTE, -1f));
                        break;
                    case TeambrellaUris.TEAMMATE_CHAT: {
                        setTeammateVoteValue(votingPart.getFloat(TeambrellaModel.ATTR_DATA_MY_VOTE, -1f));
                    }
                    break;
                }
            } else {
                mVoteButton.setVisibility(View.GONE);
            }
        }

        if (mLastRead != null && mLastRead >= 0) {
            IDataPager<JsonArray> pager = mDataHost.getPager(mTag);

            int moveTo = pager.getLoadedData().size() - 1;
            for (int i = 0; i < pager.getLoadedData().size(); i++) {
                JsonWrapper item = new JsonWrapper(pager.getLoadedData().get(i).getAsJsonObject());
                long created = item.getLong(TeambrellaModel.ATTR_DATA_CREATED, -1);
                if (created >= mLastRead) {
                    moveTo = i;
                    break;
                }
            }
            LinearLayoutManager manager = (LinearLayoutManager) mList.getLayoutManager();
            manager.scrollToPositionWithOffset(moveTo, 0);
            mLastRead = -1L;
        }
    }


    private void setClaimVoteValue(float value) {
        if (value >= 0) {
            mVoteValueView.setText(Html.fromHtml(getString(R.string.vote_in_percent_format_string, (int) (value * 100))));
        } else {
            mVoteValueView.setText(R.string.no_teammate_vote_value);
        }
    }

    private void setTeammateVoteValue(float value) {
        if (value >= 0) {
            mVoteValueView.setText(String.format(Locale.US, "%.2f", value));
        } else {
            mVoteValueView.setText(R.string.no_teammate_vote_value);
        }
    }

    private void onTeammateClickListener(View v) {
        switch (v.getId()) {
            case R.id.voting_panel:
            case R.id.vote:
                startActivityForResult(TeammateActivity.getIntent(getContext(), mDataHost.getTeamId(), mDataHost.getUserId(), mUserName, mDataHost.getImageUri()), 10);
                break;
        }
    }

    private void onClaimClickListener(View v) {
        switch (v.getId()) {
            case R.id.voting_panel:
            case R.id.vote:
                startActivityForResult(ClaimActivity.getLaunchIntent(getContext(), mDataHost.getClaimId(), mDataHost.getObjectName(), mDataHost.getTeamId()), 10);
                break;
        }
    }


}

File: app/src/main/java/com/teambrella/android/ui/chat/ChatFragment.kt
package com.teambrella.android.ui.chat

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ObjectAnimator
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.CoordinatorLayout
import android.support.v7.widget.LinearLayoutManager
import android.text.Html
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.api.server.TeambrellaUris
import com.teambrella.android.ui.base.ADataPagerProgressFragment
import com.teambrella.android.ui.base.ATeambrellaDataPagerAdapter
import com.teambrella.android.ui.claim.ClaimActivity
import com.teambrella.android.ui.claim.MODE_CHAT
import com.teambrella.android.ui.teammate.getTeammateIntent
import com.teambrella.android.ui.util.setAvatar
import com.teambrella.android.ui.util.setImage
import com.teambrella.android.util.AmountCurrencyUtil
import io.reactivex.Notification
import java.text.DecimalFormat
import java.util.*

class KChatFragment : ADataPagerProgressFragment<IChatActivity>() {


    private companion object {
        const val VOTING_FRAGMENT_TAG = "voting_fragment_tag"

        private fun fadeIn(view: View?) {
            view?.let { _view ->
                val animator = ObjectAnimator.ofFloat(_view, "alpha", 0f, 1f)
                animator.duration = 300
                _view.visibility = View.VISIBLE
                animator.start()
            }

        }

        private fun fadeOut(view: View?) {
            view?.let { _view ->
                val animator = ObjectAnimator.ofFloat(_view, "alpha", 1f, 0f)
                animator.duration = 300
                animator.addListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        super.onAnimationEnd(animation)
                        _view.visibility = View.INVISIBLE
                    }
                })
                animator.start()
            }
        }
    }


    private val votingPanelView: View? by ViewHolder(R.id.voting_panel)
    private val titleView: TextView? by ViewHolder(R.id.title)
    private val subtitleView: TextView? by ViewHolder(R.id.subtitle)
    private val voteValueView: TextView? by ViewHolder(R.id.vote_value)
    private val voteButtonView: View? by ViewHolder(R.id.vote)
    private val voteAction: View? by ViewHolder(R.id.vote_action)
    private val voteTitleView: TextView? by ViewHolder(R.id.vote_title)
    private val iconView: ImageView? by ViewHolder(R.id.image)
    private val votingContainerView: View? by ViewHolder(R.id.voting_container)
    private val votingSectionView: View? by ViewHolder(R.id.voting_section)
    private val dividerView: View? by ViewHolder(R.id.divider)
    private val hideButtonView: View? by ViewHolder(R.id.hide)

    private val decimalFormat = DecimalFormat.getInstance()


    private val votingContainerBehaviour = VotingContainerBehaviour(object : AVotingViewBehaviour.OnHideShowListener {
        override fun onHide() {
            voteButtonView?.visibility = View.VISIBLE
            hideButtonView?.visibility = View.INVISIBLE
            fadeIn(votingSectionView)
            fadeIn(dividerView)
        }

        override fun onShow() {
            voteButtonView?.visibility = View.INVISIBLE
            hideButtonView?.visibility = View.VISIBLE
            fadeOut(votingSectionView)
            fadeOut(dividerView)
        }
    })


    private val votingPanelBehaviour = VotingPanelBehaviour()


    private var userName: String? = null
    private var lastRead: Long? = null

    val layoutManager; get() = list?.layoutManager as LinearLayoutManager
    val scrollY; get() = layoutManager.findFirstVisibleItemPosition()
    val scrollYOffset: Int
        get() {
            val v = layoutManager.getChildAt(0)
            val top = if (v == null) 0 else (v.top - layoutManager.paddingTop)
            return top
        }

    override fun createAdapter(): ATeambrellaDataPagerAdapter {
        return KChatAdapter(dataHost.getPager(tags[0]), context!!, dataHost.teamId,
                when (TeambrellaUris.sUriMatcher.match(dataHost.chatUri)) {
                    TeambrellaUris.CLAIMS_CHAT -> MODE_CLAIM
                    TeambrellaUris.TEAMMATE_CHAT -> MODE_APPLICATION
                    TeambrellaUris.CONVERSATION_CHAT -> MODE_CONVERSATION
                    else -> MODE_DISCUSSION
                })
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        when (TeambrellaUris.sUriMatcher.match(dataHost.chatUri)) {
            TeambrellaUris.CLAIMS_CHAT ->
                dataHost.getObservable(ChatActivity.VOTE_DATA_TAG).observe(lifecycleOwner,
                        android.arch.lifecycle.Observer { it?.let { onChatDataUpdated(it) } })
        }

        super.onViewCreated(view, savedInstanceState)
        isRefreshable = false
        list?.setBackgroundColor(Color.TRANSPARENT)
        layoutManager.stackFromEnd = true


        when (TeambrellaUris.sUriMatcher.match(dataHost.chatUri)) {
            TeambrellaUris.CLAIMS_CHAT -> {
                votingPanelView?.visibility = View.VISIBLE
                votingPanelView?.setOnClickListener(this::onClaimClickListener)
                voteButtonView?.setOnClickListener(this::onClaimClickListener)
                hideButtonView?.setOnClickListener(this::onClaimClickListener)
                childFragmentManager.apply {
                    if (findFragmentByTag(VOTING_FRAGMENT_TAG) == null) {
                        beginTransaction().add(R.id.voting_container, com.teambrella.android.ui.claim.getInstance(
                                arrayOf(ChatActivity.CLAIM_DATA_TAG
                                        , ChatActivity.VOTE_DATA_TAG), MODE_CHAT)
                                , VOTING_FRAGMENT_TAG)
                                .commit()
                    }
                }
                dataHost.load(ChatActivity.CLAIM_DATA_TAG)
            }
            TeambrellaUris.TEAMMATE_CHAT -> {
                votingPanelView?.visibility = View.VISIBLE
                votingPanelView?.setOnClickListener(this::onTeammateClickListener)
                voteButtonView?.setOnClickListener(this::onTeammateClickListener)
                votingContainerView?.visibility = View.GONE
            }
            TeambrellaUris.CONVERSATION_CHAT -> {
                votingPanelView?.visibility = View.GONE
                list?.apply {
                    setPadding(paddingLeft, 0, paddingRight, paddingBottom)
                }
                votingContainerView?.visibility = View.GONE
            }
            TeambrellaUris.FEED_CHAT -> {
                votingPanelView?.visibility = View.GONE
                list?.apply {
                    setPadding(paddingLeft, 0, paddingRight, paddingBottom)
                }
                votingContainerView?.visibility = View.GONE
            }
        }

        list?.itemAnimator = null


        votingContainerView?.layoutParams = votingContainerView?.layoutParams?.asCoordinatorParams
                ?.apply {
                    behavior = votingContainerBehaviour
                }

        votingPanelView?.layoutParams = votingPanelView?.layoutParams.asCoordinatorParams
                ?.apply {
                    behavior = votingPanelBehaviour
                }


    }

    override val contentLayout = R.layout.fragment_chat


    override fun onDataUpdated(notification: Notification<JsonObject>) {
        super.onDataUpdated(notification)
        if (notification.isOnNext) {
            val uriString = notification.value.status?.uri
            val uri = if (uriString != null) Uri.parse(uriString) else null
            if (uri != null) {
                when (TeambrellaUris.sUriMatcher.match(Uri.parse(uriString))) {
                    TeambrellaUris.CLAIMS_CHAT,
                    TeambrellaUris.FEED_CHAT,
                    TeambrellaUris.CONVERSATION_CHAT,
                    TeambrellaUris.TEAMMATE_CHAT -> onChatDataUpdated(notification)

                    TeambrellaUris.SET_CLAIM_VOTE -> onVoteDataUpdated(notification)
                }
            }
            val scroll = notification.value.data?.scroll
            if (scroll != null) {
                if (scroll >= 0) {
                    layoutManager.scrollToPositionWithOffset(scroll, notification.value.data?.scrollOffset!!)
                } else {
                    if (!scrollToLastRead()) {
                        layoutManager.scrollToPosition((adapter?.itemCount ?: 0) - 1)
                    }
                }
            }
        }
    }

    
    public fun forceUpdate() {
        list?.recycledViewPool?.clear();
        adapter?.notifyDataSetChanged();
    }
    
    
    private fun onChatDataUpdated(notification: Notification<JsonObject>) {
        if (notification.isOnNext) {
            val data = notification.value.data
            val metadata = notification.value.metadata
            var needScrollDown = metadata?.reload == true && metadata.size ?: 0 > 0
            if (metadata?.forced == true && metadata.size ?: 0 > 0) {
                val lastItemVisible = (list?.layoutManager as? LinearLayoutManager)?.findLastVisibleItemPosition()
                if (lastItemVisible == (adapter?.itemCount ?: 0) - 2) {
                    needScrollDown = true
                }
            }
            if (needScrollDown) {
                list?.layoutManager?.scrollToPosition((adapter?.itemCount ?: 0) - 1)
            }

            if (lastRead == null) {
                lastRead = data?.discussionPart?.lastRead ?: -1L
            }


            val basicPart = data?.basic
            val votingPart = data?.voting

            if (basicPart != null) {

                when (TeambrellaUris.sUriMatcher.match(dataHost.chatUri)) {
                    TeambrellaUris.CLAIMS_CHAT -> {
                        iconView?.setImage(imageLoader.getImageUrl(basicPart.smallPhoto), R.dimen.rounded_corners_4dp, R.drawable.picture_background_round_4dp)
                        titleView?.text = basicPart.model

                        subtitleView?.text = Html.fromHtml(getString(R.string.claim_amount_format_string, decimalFormat.format(Math.round(basicPart.claimAmount
                                ?: 0f)), AmountCurrencyUtil.getLocalizedCurrency(context, data.teamPart?.currency
                                ?: "")))


                        if (votingPart == null) {
                            val votingCrypto = basicPart.votingResCrypto
                            val paymentCrypto: Double? = basicPart.paymentResCrypto

                            if (votingCrypto != null && paymentCrypto != null && votingCrypto != 0.0) {
                                if ((paymentCrypto / votingCrypto).toFloat() < 0.995) {
                                    voteTitleView?.setText(R.string.being_paid_title)
                                    setClaimVoteValue((paymentCrypto / votingCrypto).toFloat())
                                } else {
                                    voteTitleView?.setText(R.string.paid_title)
                                    setClaimVoteValue(1f)
                                }
                            } else {
                                voteTitleView?.setText(R.string.team_vote)
                                setClaimVoteValue(basicPart.reimbursement ?: -1f)
                            }

                        }
                    }
                    TeambrellaUris.TEAMMATE_CHAT -> {
                        iconView?.setAvatar(imageLoader.getImageUrl(basicPart.avatar))
                        userName = basicPart.name
                        titleView?.text = userName
                        subtitleView?.text = getString(R.string.object_format_string, basicPart.model, basicPart.year)
                        subtitleView?.setAllCaps(true)

                        if (votingPart == null) {
                            voteTitleView?.setText(R.string.risk)
                            setTeammateVoteValue(basicPart.risk ?: 0f)
                        }
                    }
                }
            }

            if (votingPart != null) {
                when (TeambrellaUris.sUriMatcher.match(dataHost.chatUri)) {
                    TeambrellaUris.CLAIMS_CHAT -> setClaimVoteValue(votingPart.myVote
                            ?: votingPart.ratioVoted?.takeIf { !votingPart.canVote } ?: -1f)
                    TeambrellaUris.TEAMMATE_CHAT -> {
                        setTeammateVoteValue(votingPart.myVote
                                ?: votingPart.ratioVoted.takeIf { !votingPart.canVote } ?: -1f)
                    }
                }

                val proxyName = votingPart.proxyName

                if (votingPart.canVote) {
                    voteTitleView?.setText(if (proxyName != null) R.string.proxy_vote_title else R.string.your_vote)
                    voteAction?.visibility = View.VISIBLE
                } else {
                    voteTitleView?.setText(if (proxyName != null) R.string.proxy_vote_title else R.string.team_vote)
                    voteAction?.visibility = View.GONE
                }

            } else {
                voteAction?.visibility = View.GONE
            }
        }
    
        scrollToLastRead()
    }

    
    private fun scrollToLastRead(): Boolean {
        if (lastRead != null && lastRead!! >= 0L) {
            val pager = dataHost.getPager(tags[0])
        
            var moveTo = pager.loadedData.size() - 1
            for (i in 0 until pager.loadedData.size()) {
                val item = pager.loadedData.get(i).asJsonObject
                val created = item.created ?: -1
                if (created >= lastRead!!) {
                    moveTo = i
                    break
                }
            }
            layoutManager.scrollToPositionWithOffset(moveTo, 0)
            lastRead = -1L
            return true
        }
        return false
    }
    
    private fun onVoteDataUpdated(notification: Notification<JsonObject>) {
        setClaimVoteValue(notification.value?.data?.voting?.myVote ?: -1f)
    }


    private fun onTeammateClickListener(v: View) {
        when (v.id) {
            R.id.voting_panel, R.id.vote ->
                startActivityForResult(getTeammateIntent(context!!, dataHost.teamId, dataHost.userId, userName, dataHost.imageUri), 10)
        }
    }

    private fun onClaimClickListener(v: View) {
        when (v.id) {
            R.id.voting_panel -> startActivityForResult(
                    ClaimActivity.getLaunchIntent(v.context, dataHost.claimId, dataHost.objectName, dataHost.teamId), 10
            )
            R.id.vote -> votingContainerBehaviour.show(votingContainerView)
            R.id.hide -> votingContainerBehaviour.hide(votingContainerView)
        }
    }

    private fun setClaimVoteValue(value: Float) {
        if (value >= 0) {
            voteValueView?.text = Html.fromHtml(getString(R.string.vote_in_percent_format_string, (value * 100).toInt()))
        } else {
            voteValueView?.setText(R.string.no_teammate_vote_value)
        }
    }

    private fun setTeammateVoteValue(value: Float) {
        if (value >= 0) {
            voteValueView?.text = String.format(Locale.US, "%.2f", value)
        } else {
            voteValueView?.setText(R.string.no_teammate_vote_value)
        }
    }


    private val ViewGroup.LayoutParams?.asCoordinatorParams: CoordinatorLayout.LayoutParams?
        get() = this as CoordinatorLayout.LayoutParams


}
